File: backend/routes/filters.js
/**
 * Filter Routes
 * GET /api/filters/branches
 * GET /api/filters/categories
 * GET /api/filters/seat-types
 * GET /api/filters/college-names
 */

import express from 'express';
import { LookupData, Branch, Category, SeatType } from '../models/index.js';

const router = express.Router();

/**
 * GET /api/filters/branches
 */
router.get('/branches', async (req, res) => {
  try {
    const branches = await Branch.find({}, 'branchName -_id').lean();
    res.json({ branches: branches.map(b => ({ id: b.branchName, name: b.branchName })) });
  } catch (error) {
    res.status(500).json({ error: error.message });
  }
});

/**
 * GET /api/filters/categories
 */
router.get('/categories', async (req, res) => {
  try {
    const categories = await Category.find({}, 'categoryType displayName -_id').lean();
    res.json({ categories: categories.map(c => ({ id: c.categoryType, name: c.displayName })) });
  } catch (error) {
    res.status(500).json({ error: error.message });
  }
});

/**
 * GET /api/filters/seat-types
 */
router.get('/seat-types', async (req, res) => {
  try {
    const seatTypes = await SeatType.find({}, 'seatTypeId displayName -_id').lean();
    res.json({ seatTypes: seatTypes.map(s => ({ id: s.seatTypeId, name: s.displayName })) });
  } catch (error) {
    res.status(500).json({ error: error.message });
  }
});

/**
 * GET /api/filters/college-names
 */
router.get('/college-names', async (req, res) => {
  try {
    const lookup = await LookupData.findOne({}, 'collegeNames -_id').lean();
    res.json({ collegeNames: lookup?.collegeNames || [] });
  } catch (error) {
    res.status(500).json({ error: error.message });
  }
});

export default router;

File: api/routes/filters.py
"""
GET /api/filters/branches
GET /api/filters/categories
GET /api/filters/seat-types
GET /api/filters/all
"""

from fastapi import APIRouter, HTTPException
import traceback
from phase4_prediction.predictor import get_lookup

router = APIRouter()


@router.get("/branches")
def get_branches():
    """All unique branch names for dropdown filters."""
    try:
        lookup = get_lookup()
        if not lookup or "branches" not in lookup:
            raise ValueError("Branches data not found in lookup")
        return {"branches": lookup["branches"]}
    except Exception as e:
        traceback.print_exc()
        raise HTTPException(status_code=500, detail=f"Error fetching branches: {str(e)}")


@router.get("/categories")
def get_categories():
    """All unique category codes with their group mappings."""
    try:
        lookup = get_lookup()
        if not lookup:
            raise ValueError("Lookup data not found")
        return {
            "categories":       lookup.get("categories", []),
            "category_groups":  lookup.get("category_groups", {}),
        }
    except Exception as e:
        traceback.print_exc()
        raise HTTPException(status_code=500, detail=f"Error fetching categories: {str(e)}")


@router.get("/seat-types")
def get_seat_types():
    """All seat type options."""
    try:
        lookup = get_lookup()
        if not lookup or "seat_types" not in lookup:
            raise ValueError("Seat types data not found in lookup")
        return {"seat_types": lookup["seat_types"]}
    except Exception as e:
        traceback.print_exc()
        raise HTTPException(status_code=500, detail=f"Error fetching seat types: {str(e)}")


@router.get("/all")
def get_all_filters():
    """All filter options in one call — use this on app load."""
    try:
        lookup = get_lookup()
        if not lookup:
            raise ValueError("Lookup data not found")
        return {
            "branches":        lookup.get("branches", []),
            "categories":      lookup.get("categories", []),
            "category_groups": lookup.get("category_groups", {}),
            "seat_types":      lookup.get("seat_types", []),
        }
    except Exception as e:
        traceback.print_exc()
        raise HTTPException(status_code=500, detail=f"Error fetching filters: {str(e)}")
